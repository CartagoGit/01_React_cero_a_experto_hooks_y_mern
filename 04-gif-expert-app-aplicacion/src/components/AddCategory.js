import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ categories="", setCategories }) => {
	const [inputValue , setInputValue] = useState(categories);
	// console.log(inputValue);

	const handleInputChange = (e) => {
		console.log("-> handleInputChange");
		setInputValue(e.target.value);
	};

	const handleSubmit = (e) => {
		console.log("-> handleSubmit",inputValue);
		e.preventDefault();

		if (inputValue.trim().length > 0) {
			// setCategories((c) => [inputValue, ...c]);
			setCategories(() => inputValue);
			setInputValue("");
		} else setInputValue("");
	};

	return (
		<form onSubmit={handleSubmit}>
			<input type='text' value={inputValue} onChange={handleInputChange} />
		</form>
	);
};

AddCategory.propTypes = {
	setCategories: PropTypes.func.isRequired
};
