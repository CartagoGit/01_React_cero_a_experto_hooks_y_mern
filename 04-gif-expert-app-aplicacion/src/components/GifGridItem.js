import React from 'react'

export const GifGridItem = ({id, title, url}) => {
    // console.log(id,title,url);
    return (
        <div className='card animate__tada' id={id}>
            {/* {props} */}
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
}
