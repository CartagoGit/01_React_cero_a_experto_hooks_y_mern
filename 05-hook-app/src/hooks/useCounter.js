import { useState } from "react";

export const useCounter = (initialState = 0) => {
	const [counter, setCounter] = useState(initialState);

	const increment = (num=1) => {
		setCounter(counter => counter + num);
	};
	const decrement = (num=1) => {
		setCounter(counter => counter - num);
	};
	const reset = () => {
		setCounter(counter => counter = initialState);
	};

	return {
		counter,
		increment,
		decrement,
		reset,
		setCounter
	};
};
