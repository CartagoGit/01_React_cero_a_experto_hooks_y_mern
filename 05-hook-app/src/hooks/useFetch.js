import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
	const isMounted = useRef(true);
	const [state, setState] = useState({
		data: null,
		loading: true,
		error: null
	});

	useEffect(() => {
		return () => {
			isMounted.current = false;
		};
	}, []);

	useEffect(() => {
		setState({ data: null, loading: true, error: null });
		fetch(url)
			.then((resp) => resp.json())
			.then((data) => {
				// aqui validamos el arreglo no este en cero
				// console.log(url);
				// if (data.length === 0)
				// 	data.push({
				// 		quote: "Without Data"
				// 	});
				if (data.length === 0) throw new EmptyStackException();
				if (isMounted.current) {
					setState({
						loading: false,
						error: null,
						data
					});
				}
			})
			.catch((error) =>
				setState({
					loading: false,
					error: 'Error fetching url',
					data: null
				})
			);
	}, [url]);

	return state;
};
