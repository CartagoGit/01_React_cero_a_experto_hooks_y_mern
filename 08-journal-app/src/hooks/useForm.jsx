import { useState } from "react";

export const useForm = (initialState = {}) => {
	const [values, setValues] = useState(initialState);	const reset = (newFormState = initialState) => {
		setValues(() => newFormState);
	};

	const handleChangeValue = ({ target }) => {
		setValues({
			...values,
			[target.name]: target.value
		});
	};

	return [values, handleChangeValue, reset];
};
