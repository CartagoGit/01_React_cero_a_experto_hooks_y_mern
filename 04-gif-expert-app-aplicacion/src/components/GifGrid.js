// import React, { useState, useEffect } from "react";
import React from "react";
import { useFetchGif } from "../hooks/useFetchGifs";
// import { getGifs } from "../helpers/getGifs";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
	// const [count, setCount] = useState(0);
	// const [images, setImages] = useState([]);

	// useEffect(() => {
	// 	getGifs(category).then((imgs) => setImages(imgs));
	// 	//// eslint-disable-next-line react-hooks/exhaustive-deps
	// }, [category]);

	const { data: images, loading } = useFetchGif(category);

	return (
		<>
			<h3>{category}</h3>
            {loading && <p>Cargando...</p>}

			

			<div className='cardGrid'>
				{/* { <h3>{count}</h3> } */}
				{/* <button onClick={ () => setCount(count+1)}></button> */}

				{images.map((img) => (
					// <li key={id}> {title}</li>
					// <GifGridItem key={img.id} img ={img}/>
					<GifGridItem key={img.id} {...img} />
				))}
			</div>
		</>
	);
};
