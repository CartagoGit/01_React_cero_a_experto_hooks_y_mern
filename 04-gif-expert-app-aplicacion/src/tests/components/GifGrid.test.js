import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock("../../hooks/useFetchGifs");

describe("Pruebas en GifGrid.js", () => {
	let wrapper;

	const fillWrapper = (cat = "un string", dat = [], load = true) => {
		useFetchGifs.mockReturnValue({
			data: dat,
			loading: load
		});
		wrapper = shallow(<GifGrid category={cat} />);
	};
	beforeEach(() => fillWrapper());

	test("Deberia comparar el snapshoot ", () => {
		console.log("------> " + expect.getState().currentTestName + " <------");

		expect(wrapper).toMatchSnapshot();
	});

	test("Deberia mostrar items cuando se cargan imagenes con el hook useFetchGifs ", () => {
		console.log("------> " + expect.getState().currentTestName + " <------");

		const gifs = [
			{ id: "ABC", url: "https://cualquiercosa.jpg", title: "Cualquier titulo" },
            { id: "CBA", url: "https://cualquiercosa.jpg", title: "Cualquier titulo" }
		];
		fillWrapper("una categoria cualquiera", gifs, false);

        expect(wrapper).toMatchSnapshot();
		expect(wrapper.find("p").exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
	});
});
