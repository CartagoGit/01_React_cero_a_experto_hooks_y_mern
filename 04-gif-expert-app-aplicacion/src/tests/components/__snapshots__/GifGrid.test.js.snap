// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifGrid.js Deberia comparar el snapshoot  1`] = `
<Fragment>
  <h3>
    un string
  </h3>
  <p>
    Cargando...
  </p>
  <div
    className="cardGrid"
  />
</Fragment>
`;

exports[`Pruebas en GifGrid.js Deberia mostrar items cuando se cargan imagenes con el hook useFetchGifs  1`] = `
<Fragment>
  <h3>
    una categoria cualquiera
  </h3>
  <div
    className="cardGrid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier titulo"
      url="https://cualquiercosa.jpg"
    />
    <GifGridItem
      id="CBA"
      key="CBA"
      title="Cualquier titulo"
      url="https://cualquiercosa.jpg"
    />
  </div>
</Fragment>
`;
